File: Napakalaki_Ruby/pkg/Napakalaki_Ruby-0.0.1/lib/card_dealer.rb
#encoding: utf-8

=begin
  Marco Manuel Fernández Pranno, 2015 (Granada)
  Para la asignatura PDOO (Universidad de Granada)
  Licencia GPLv3 a partir del 15/02/16, hasta entonces todos los derechos reservados.
=end 

require 'singleton'
require_relative 'treasure'
require_relative 'cultist'
require_relative 'cultist_player'
require_relative 'specific_bad_consequence'
require_relative 'numeric_bad_consequence'
require_relative 'death_bad_consequence'

class CardDealer
  include Singleton
  
  def initialize()
    @usedMonsters = Array.new
    @unusedMonsters = Array.new
    
    @usedTreasures = Array.new
    @unusedTreasures = Array.new
    
    @unusedCultists = Array.new
    @usedCultists = Array.new
    
  end
  
  def initCultistCardDeck()
    @unusedCultists << Cultist.new("Cultist",1)
    @unusedCultists << Cultist.new("Cultist",1)
    @unusedCultists << Cultist.new("Cultist",1)
    @unusedCultists << Cultist.new("Cultist",1)
    @unusedCultists << Cultist.new("Cultist",2)
    @unusedCultists << Cultist.new("Cultist",2)
  end
  
  def initTreasureCardDeck()
    @unusedTreasures << Treasure.new("¡Sí mi amo!", 4, TreasureKind::HELMET)
    @unusedTreasures << Treasure.new("Botas de investigación", 3, TreasureKind::SHOES)
    @unusedTreasures << Treasure.new("Capucha de Cthulhu", 3, TreasureKind::HELMET)
    @unusedTreasures << Treasure.new("A prueba de balas", 2, TreasureKind::ARMOR)
    @unusedTreasures << Treasure.new("Botas de lluvia ácida", 1, TreasureKind::BOTHHANDS) # ¿ BOTHHANDS ?
    @unusedTreasures << Treasure.new("Casco minero", 2, TreasureKind::HELMET)
    @unusedTreasures << Treasure.new("Ametralladora Thompson", 4, TreasureKind::BOTHHANDS)
    @unusedTreasures << Treasure.new("Camiseta de la UGR", 1, TreasureKind::ARMOR)
    @unusedTreasures << Treasure.new("Clavo de rail ferroviario", 3, TreasureKind::ONEHAND)
    @unusedTreasures << Treasure.new("Cuchillo de sushi arcano", 2, TreasureKind::ONEHAND)
    @unusedTreasures << Treasure.new("Fez alópodo", 3, TreasureKind::HELMET)
    @unusedTreasures << Treasure.new("Hacha prehistórica", 2, TreasureKind::ONEHAND)
    @unusedTreasures << Treasure.new("El aparato del Pr. Tesla", 4, TreasureKind::ARMOR)
    @unusedTreasures << Treasure.new("Gaita", 4, TreasureKind::BOTHHANDS)
    @unusedTreasures << Treasure.new("Insecticida", 2, TreasureKind::ONEHAND)
    @unusedTreasures << Treasure.new("Escopeta de 3 cañones", 4, TreasureKind::BOTHHANDS)
    @unusedTreasures << Treasure.new("Garabato místico", 2, TreasureKind::ONEHAND)
    @unusedTreasures << Treasure.new("La rebeca metálica", 2, TreasureKind::ARMOR)
    @unusedTreasures << Treasure.new("Lanzallamas", 4, TreasureKind::BOTHHANDS)
    @unusedTreasures << Treasure.new("Necrocomicón", 1, TreasureKind::ONEHAND)
    @unusedTreasures << Treasure.new("Necronomicón", 5, TreasureKind::BOTHHANDS)
    @unusedTreasures << Treasure.new("Linterna a 2 manos", 3, TreasureKind::BOTHHANDS)
    @unusedTreasures << Treasure.new("Necrognomicón", 2, TreasureKind::ONEHAND)
    @unusedTreasures << Treasure.new("Necrotelecom", 2, TreasureKind::HELMET)
    @unusedTreasures << Treasure.new("Mazo de los antiguos", 3, TreasureKind::ONEHAND)
    @unusedTreasures << Treasure.new("Necroplayboycón", 3, TreasureKind::ONEHAND)
    @unusedTreasures << Treasure.new("Porra prenatural", 2, TreasureKind::ONEHAND)
    @unusedTreasures << Treasure.new("Shogulador", 1, TreasureKind::BOTHHANDS)
    @unusedTreasures << Treasure.new("Varita de atizamiento", 3, TreasureKind::ONEHAND)
    @unusedTreasures << Treasure.new("Tentáculo de pega", 2, TreasureKind::HELMET)
    @unusedTreasures << Treasure.new("Zapato deja-amigos", 1, TreasureKind::SHOES)
    
  end
  
  def initMonsterCardDeck()

    prize = Prize.new(2,1)
    bad_consequence = SpecificBadConsequence.new('Pierdes tu armadura visible y otra oculta', 0, [TreasureKind::ARMOR], [TreasureKind::ARMOR])
    @unusedMonsters << Monster.new('3 Byakhees de bonanza',8,prize,bad_consequence)

    prize = Prize.new(1,1)
    bad_consequence = SpecificBadConsequence.new('Embobados con el lindo primigenio te descartas de tu casco visible.', 0, [TreasureKind::HELMET], nil)
    @unusedMonsters << Monster.new('Chibithulhu',2,prize,bad_consequence)

    prize = Prize.new(1,1)
    bad_consequence = SpecificBadConsequence.new('El primordial bostezo contagioso. Pierdes el calzado visible.', 0, [TreasureKind::SHOES],nil)
    @unusedMonsters << Monster.new('El sopor de Dunwich',2,prize,bad_consequence)

    prize = Prize.new(4,1)
    bad_consequence = SpecificBadConsequence.new('Te atrapan para llevarte de fiesta y te dejan caer en mitad del vuelo. Descarta 1 mano visible y 1 mano oculta',0,
    [TreasureKind::ONEHAND],[TreasureKind::ONEHAND])
    @unusedMonsters << Monster.new('Ángeles de la noche ibicenca',14,prize,bad_consequence)

    prize = Prize.new(3,1)
    bad_consequence = NumericBadConsequence.new('Pierdes todos tus tesoros visibles.', 0, BadConsequence::MAXTREASURES, 0)
    @unusedMonsters << Monster.new('El gorrón en el umbral', 10, prize, bad_consequence)

    prize = Prize.new(2,1)
    bad_consequence = SpecificBadConsequence.new('Pierdes la armadura visible', 0, [TreasureKind::ARMOR],nil)
    @unusedMonsters << Monster.new('H.P Munchcraft',6,prize,bad_consequence)

    prize = Prize.new(1,1)
    bad_consequence = SpecificBadConsequence.new('Sientes bichos bajo la ropa. Descarta la armadura visible.', 0, [TreasureKind::ARMOR],nil)
    @unusedMonsters << Monster.new('Bichgooth',2,prize,bad_consequence)

    prize = Prize.new(4,2)
    bad_consequence = NumericBadConsequence.new('Pierdes 5 niveles y 3 tesoros visibles', 5, 3, 0)
    @unusedMonsters << Monster.new('El rey de rosa', 13, prize, bad_consequence)

    prize = Prize.new(1,1)
    bad_consequence = NumericBadConsequence.new('Toses los pulmones y pierdes 2 niveles.', 2, 0,0)
    @unusedMonsters << Monster.new('La que redacta en las tinieblas.',2,prize,bad_consequence)

    prize = Prize.new(2,1)
    bad_consequence = DeathBadConsequence.new('Estos monstruos resultan bastante superficiales y te aburren mortalmente. Estás muerto.')
    @unusedMonsters << Monster.new('Los hondos',8, prize, bad_consequence)

    prize = Prize.new(2,1)
    bad_consequence = NumericBadConsequence.new('Pierdes 2 niveles y 2 tesoros ocultos.',2,0,2)
    @unusedMonsters << Monster.new('Semillas Cthulhu', 4, prize, bad_consequence)

    prize = Prize.new(2,1)
    bad_consequence = SpecificBadConsequence.new('Te intentas escaquear. Pierdes una mano visible.',0, [TreasureKind::ONEHAND],nil)
    @unusedMonsters << Monster.new('Dameargo', 1, prize, bad_consequence)

    prize = Prize.new(1,1)
    bad_consequence = NumericBadConsequence.new('Da mucho asquito. Pierdes 3 niveles.', 3, 0,0)
    @unusedMonsters << Monster.new('Pollipólipo volante',3,prize,bad_consequence)

    prize = Prize.new(3,1)
    bad_consequence = DeathBadConsequence.new('No lo hace mucha gracia que pronuncien mal su nombre. Estás muerto.')
    @unusedMonsters << Monster.new('Yskhtihyssg-Goth',12,prize,bad_consequence)

    prize = Prize.new(4,1)
    bad_consequence = DeathBadConsequence.new('La familia te atrapa. Estás muerto.')
    @unusedMonsters << Monster.new('Familia Feliz',1,prize,bad_consequence)

    prize = Prize.new(2,1)
    bad_consequence = SpecificBadConsequence.new('La quinta directiva primaria te obliga a perder 2 niveles y un tesoro de ambas manos visible.', 
      2, [TreasureKind::BOTHHANDS],nil)
    @unusedMonsters << Monster.new('Roboggoth',8,prize,bad_consequence)

    prize = Prize.new(1,1)
    bad_consequence = SpecificBadConsequence.new('Te asusta en la noche. Pierdes un casco visible.',0,[TreasureKind::HELMET],nil)
    @unusedMonsters << Monster.new('El espía',5,prize,bad_consequence)

    prize = Prize.new(1,1)
    bad_consequence = NumericBadConsequence.new('Menudo susto te llevas. Pierdes 2 niveles y 5 tesoros visibles.', 2, 5,0)
    @unusedMonsters << Monster.new('El lenguas',20,prize,bad_consequence)

    prize = Prize.new(1,1)
    bad_consequence = SpecificBadConsequence.new('Te faltan manos para tanta cabeza. Pierdes 3 niveles y tus tesoros visibles de las dos manos.',
      3, [TreasureKind::BOTHHANDS],nil)
    @unusedMonsters << Monster.new('Bicéfalo',20,prize,bad_consequence)
    
    prize = Prize.new(3,1)
    bad_consequence = SpecificBadConsequence.new("Pierdes 1 mano visible", 0, [TreasureKind::ONEHAND],nil)
    @unusedMonsters << Monster.new("El mal indecible impronunciable", 10, prize, bad_consequence, -2)
    
    prize = Prize.new(2,1)
    bad_consequence = NumericBadConsequence.new("Pierdes tus tesoros visibles. Jajaja.",0,BadConsequence::MAXTREASURES, 0)
    @unusedMonsters << Monster.new("Testigos Oculares",6,prize,bad_consequence,+2)
    
    prize = Prize.new(2,5)
    bad_consequence = DeathBadConsequence.new("Hoy no es tu día de suerte. Mueres.")
    @unusedMonsters << Monster.new("El gran cthulhu",20,prize,bad_consequence,+4)
    
    prize = Prize.new(2,1)
    bad_consequence = NumericBadConsequence.new("Tu gobierno te recorta 2 niveles.", 2, 0,0)
    @unusedMonsters << Monster.new("Serpiente Político",8,prize,bad_consequence,-2)
    
    prize = Prize.new(1,1)
    bad_consequence = SpecificBadConsequence.new("Pierdes tu casco y tu armadura visible. Pierdes tus manos ocultas.",0,[TreasureKind::ARMOR,TreasureKind::HELMET],[TreasureKind::BOTHHANDS])
    @unusedMonsters << Monster.new("Felpuggoth",2,prize,bad_consequence,+5)
    
    prize = Prize.new(4,2)
    bad_consequence = NumericBadConsequence.new("Pierdes 2 niveles",2,0,0)
    @unusedMonsters << Monster.new("Shoggoth",16,prize,bad_consequence,-4)
    
    prize = Prize.new(1,1)
    bad_consequence = NumericBadConsequence.new("Pintalabios negro. Pierdes 2 niveles.",2,0,0)
    @unusedMonsters << Monster.new("Lolitagooth",2,prize,bad_consequence,+3)
    
  end
  
  def shuffleCultists()
    
    @unusedCultists.each_with_index { |cultist,indice| 
      
      pos_nueva = rand(@unusedCultists.size-1)
      aux = @unusedCultists[pos_nueva]
      @unusedCultists[indice] = aux
      @unusedCultists[pos_nueva] = cultist
      
    }
  end
  
  def shuffleTreasures()
    
    @unusedTreasures.each_with_index { |tesoro,indice| 
      
      pos_nueva = rand(@unusedTreasures.size-1)
      aux = @unusedTreasures[pos_nueva]
      @unusedTreasures[indice] = aux
      @unusedTreasures[pos_nueva] = tesoro
      
    }
      
  end
  
  def shuffleMonsters()
    
    @unusedMonsters.each_with_index { |monstruo,indice| 
      
      pos_nueva = rand(@unusedMonsters.size-1)
      aux = @unusedMonsters[pos_nueva]
      @unusedMonsters[indice] = aux
      @unusedMonsters[pos_nueva] = monstruo
      
    }
  end
  
  def nextCultist()
    if(@unusedCultists.empty?)
      punteroAux = @unusedCultists
      @unusedCultists = @usedCultists
      @usedCultists = punteroAux
      shuffleCultists()
    end
    nextC = @unusedCultists[0]
    @unusedCultists.delete_at(0)
    return nextC
  end
  
  def nextTreasure()
    if(@unusedTreasures.empty?)
      punteroAux = @unusedTreasures
      @unusedTreasures = @usedTreasures
      @usedTreasures = punteroAux
      shuffleTreasures()
    end
    nextT = @unusedTreasures[0]
    @unusedTreasures.delete_at(0)
    return nextT
  end
  
  def nextMonster()
    if(@unusedMonsters.empty?)
      punteroAux = @unusedMonsters
      @unusedMonsters = @usedMonsters
      @usedMonsters = punteroAux
      shuffleMonsters()
    end
    nextM = @unusedMonsters[0]
    @unusedMonsters.delete_at(0)
    return nextM
  end
  
  def giveTreasureBack(t)
    
    @usedTreasures << t
    @unusedTreasures.delete(t)
    
  end
  
  def giveMonsterBack(m)
    
    @usedMonsters << m
    @unusedMonsters.delete(m)
    
  end
  
  def initCards()
    initTreasureCardDeck()
    initMonsterCardDeck()
    initCultistCardDeck()
  end
  
  private_class_method :new
  private :initTreasureCardDeck,:initMonsterCardDeck,:shuffleTreasures,:shuffleMonsters
end

File: Napakalaki_Java/src/NapakalakiGame/CardDealer.java
/**
 *  Marco Manuel Fernández Pranno, 2015 (Granada)
 *  Para la asignatura PDOO (Universidad de Granada)
 *  Licencia GPLv3 a partir del 15/02/16, hasta entonces todos los derechos reservados.
 */

package NapakalakiGame;

import java.util.ArrayList;
import java.util.Random;

public class CardDealer {
    
    private static final CardDealer instance = new CardDealer();
    
    
    ArrayList<Monster> unusedMonsters = null;
    ArrayList<Monster> usedMonsters = null;
    ArrayList<Treasure> unusedTreasures = null;
    ArrayList<Treasure> usedTreasures = null;
    ArrayList<Cultist> unusedCultists = null;
    ArrayList<Cultist> usedCultists = null;
    
    private CardDealer(){
        
        this.unusedMonsters = new ArrayList();
        this.usedMonsters = new ArrayList();
        this.unusedTreasures = new ArrayList();
        this.usedTreasures = new ArrayList();
        this.unusedCultists = new ArrayList();
        this.usedCultists = new ArrayList();

    }
    
    static CardDealer getInstance(){
        return instance;
    }
    
    private void initTreasureCardDeck() {
        Treasure tesoro = new Treasure("¡Sí mi amo!", 4, TreasureKind.HELMET);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("Botas de investigación", 3, TreasureKind.SHOES);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("Capucha de Cthulhu", 3, TreasureKind.HELMET);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("A prueba de balas", 2, TreasureKind.ARMOR);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("Botas de lluvia ácida", 1, TreasureKind.BOTHHANDS); // ¿ BOTHHANDS ?
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("Casco minero", 2, TreasureKind.HELMET);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("Ametralladora Thompson", 4, TreasureKind.BOTHHANDS);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("Camiseta de la UGR", 1, TreasureKind.ARMOR);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("Clavo de rail ferroviario", 3, TreasureKind.ONEHAND);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("Cuchillo de sushi arcano", 2, TreasureKind.ONEHAND);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("Fez alópodo", 3, TreasureKind.HELMET);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("Hacha prehistórica", 2, TreasureKind.ONEHAND);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("El aparato del Pr. Tesla", 4, TreasureKind.ARMOR);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("Gaita", 4, TreasureKind.BOTHHANDS);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("Insecticida", 2, TreasureKind.ONEHAND);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("Escopeta de 3 cañones", 4, TreasureKind.BOTHHANDS);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("Garabato místico", 2, TreasureKind.ONEHAND);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("La rebeca metálica", 2, TreasureKind.ARMOR);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("Lanzallamas", 4, TreasureKind.BOTHHANDS);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("Necrocomicón", 1, TreasureKind.ONEHAND);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("Necronomicón", 5, TreasureKind.BOTHHANDS);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("Linterna a 2 manos", 3, TreasureKind.BOTHHANDS);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("Necrognomicón", 2, TreasureKind.ONEHAND);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("Necrotelecom", 2, TreasureKind.HELMET);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("Mazo de los antiguos", 3, TreasureKind.ONEHAND);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("Necroplayboycón", 3, TreasureKind.ONEHAND);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("Porra prenatural", 2, TreasureKind.ONEHAND);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("Shogulador", 1, TreasureKind.BOTHHANDS);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("Varita de atizamiento", 3, TreasureKind.ONEHAND);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("Tentáculo de pega", 2, TreasureKind.HELMET);
        unusedTreasures.add(tesoro);

        tesoro = new Treasure("Zapato deja-amigos", 1, TreasureKind.SHOES);
        unusedTreasures.add(tesoro);
    }
    
    private void initCultistCardDeck(){
        unusedCultists.add(new Cultist("Cultist",1));
        unusedCultists.add(new Cultist("Cultist",1));
        unusedCultists.add(new Cultist("Cultist",1));
        unusedCultists.add(new Cultist("Cultist",1));
        unusedCultists.add(new Cultist("Cultist",2));
        unusedCultists.add(new Cultist("Cultist",2));
    }

    private void initMonsterCardDeck(){
        
        BadConsequence bc;
        Prize prize;
        ArrayList<TreasureKind> tVisible = new ArrayList();
        ArrayList<TreasureKind> tHidden = new ArrayList();
        
        // El Rey de Rosa
       bc = new NumericBadConsequence("Pierdes 5 niveles y 3 tesoros visibles.", 5, 3, 0);
       prize = new Prize(4,2);
       unusedMonsters.add(new Monster("El rey de rosa", 13, bc, prize));
       
       // Ángeles de la noche ibicenca
 
       tVisible.add(TreasureKind.ONEHAND);
       tHidden.add(TreasureKind.ONEHAND);
       bc = new SpecificBadConsequence("Tatrapan para llevarte de fiesta y te dejan caer en mitad del vuelo. "
               + "Descarta 1 mano visible y 1 mano oculta.", 0, tVisible, tHidden);
       prize = new Prize(4,1);
       unusedMonsters.add(new Monster("Angeles de la noche ibicenca", 14, bc, prize));
       
       tVisible = new ArrayList();
       tHidden = new ArrayList();
       
       // 3 Byakhees de bonanza
       tVisible.add(TreasureKind.ARMOR);
       tHidden.add(TreasureKind.ARMOR);
       bc = new SpecificBadConsequence("Pierdes tu armadura visible y otra oculta.",0,tVisible,tHidden);
       prize = new Prize(2,1);
       unusedMonsters.add(new Monster("3 Byakhees de bonanza",8,bc,prize));
       
       tVisible = new ArrayList();
       tHidden = new ArrayList();
       
       // Chibithulhu
       tVisible.add(TreasureKind.HELMET);
       bc = new SpecificBadConsequence("Embobados con el lindo primigenio te descartas de tu casco visible.",
       0,tVisible,null);
       prize = new Prize(1,1);
       unusedMonsters.add(new Monster("Chibithulhu",2,bc,prize));
       
       tVisible = new ArrayList();
       
       // El sopor de Dunwich
       tVisible.add(TreasureKind.SHOES);
       bc = new SpecificBadConsequence("El primordial bostezo contagioso. Pierdes el calzado visible.",0,tVisible,null);
       prize = new Prize(1,1);
       unusedMonsters.add(new Monster("El sopor de Dunwich",2,bc,prize));
       
       tVisible = new ArrayList();
       
       // El gorrón en el umbral
       
       bc = new NumericBadConsequence("Pierdes todos tus tesoros visibles.",0,BadConsequence.MAXTREASURES,0);
       prize = new Prize(3,1);
       unusedMonsters.add(new Monster("El gorrón en el umbral",10,bc,prize));
       
       
       // H.P Munchcraft
       
       tVisible.add(TreasureKind.ARMOR);
       bc = new SpecificBadConsequence("Pierdes la armadura visible.",0,tVisible,null);
       prize = new Prize(2,1);
       unusedMonsters.add(new Monster("H.P. Munchcraft",6,bc,prize));
       
       tVisible = new ArrayList();
       
       // Bichgooth
       
       tVisible.add(TreasureKind.ARMOR);
       bc = new SpecificBadConsequence("Sientes bichos bajo la ropa. Descarta la armadura visible.",0,tVisible,null);
       prize = new Prize(1,1);
       unusedMonsters.add(new Monster("Bichgooth",2,bc,prize));
       
       tVisible = new ArrayList();
       
       // La que redacta en las tinieblas
       
       bc = new NumericBadConsequence("Toses los pulmones y pierdes 2 niveles.", 2,0,0);
       prize = new Prize(1,1);
       unusedMonsters.add(new Monster("La que redacta en las tinieblas",2,bc,prize));
       
       // Los hondos
       
       bc = new DeathBadConsequence("Estos unusedMonsters resultan bastante superficiales y te aburren mortalmente. Estás muerto.");
       prize = new Prize(2,1);
       unusedMonsters.add(new Monster("Los hondos",8,bc,prize));
       
       // Semillas Cthulhu
       
       bc = new NumericBadConsequence("Pierdes 2 niveles y 2 tesoros ocultos.",2,0,2);
       prize = new Prize(2,1);
       unusedMonsters.add(new Monster("Semillas Cthulhu",4,bc,prize));
       
       // Dameargo
       
       tVisible.add(TreasureKind.ONEHAND);
       bc = new SpecificBadConsequence("Te intentas escaquear. Pierdes una mano visible.",0,tVisible,null);
       prize = new Prize(2,1);
       unusedMonsters.add(new Monster("Dameargo",1,bc,prize));
       
       tVisible = new ArrayList();
       
       // Pollipólipo volante
       
       bc = new NumericBadConsequence("Da mucho asquito. Pierdes 3 niveles.",3,0,0);
       prize = new Prize(1,1);
       unusedMonsters.add(new Monster("Pollipólipo volante",3,bc,prize));
       
       // Yskhtihyssg-Goth
       
       bc = new DeathBadConsequence("No le hace gracia que pronuncien mal su nombre. Estás muerto");
       prize = new Prize(3,1);
       unusedMonsters.add(new Monster("Yskhtihyssg-Goth",12,bc,prize));
       
       // Familia feliz
       
       bc = new DeathBadConsequence("La familia te atrapa. Estás muerto");
       prize = new Prize(4,1);
       unusedMonsters.add(new Monster("Familia feliz",1,bc,prize));
       
       // Roboggoth
       tVisible.add(TreasureKind.BOTHHANDS);
       bc = new SpecificBadConsequence("La quinta directiva primaria te obliga a perder 2 niveles y un tesoro de las dos manos visible.",2,tVisible,null);
       prize = new Prize(2,1);
       unusedMonsters.add(new Monster("Roboggoth",8,bc,prize));
       
       tVisible = new ArrayList();
       
       // El espia
       
       tVisible.add(TreasureKind.HELMET);
       bc = new SpecificBadConsequence("Te asusta en la noche. Pierdes un casco visible.",0,tVisible,null);
       prize = new Prize(1,1);
       unusedMonsters.add(new Monster("El espia",5,bc,prize));
       
       tVisible = new ArrayList();
       
       // El lenguas
       
       bc = new NumericBadConsequence("Menudo susto te llevas. Pierdes 2 niveles y 5 tesoros visibles.",2,5,0);
       prize = new Prize(1,1);
       unusedMonsters.add(new Monster("El lenguas",20,bc,prize));
       
       // Bicéfalo
       tVisible.add(TreasureKind.BOTHHANDS);
       bc = new SpecificBadConsequence("Te faltan manos para tanta cabeza. Pierdes 3 niveles y tus tesoros visibles de las dos manos.",3,tVisible,null);
       prize = new Prize(1,1);
       unusedMonsters.add(new Monster("Bicéfalo",20,bc,prize));
       
       tVisible = new ArrayList();
       
       tVisible.add(TreasureKind.ONEHAND);
       bc = new SpecificBadConsequence("Pierdes 1 mano visible",0,tVisible,null);
       prize = new Prize(3,1);
       unusedMonsters.add(new Monster("El mal indecible impronunciable",10,bc,prize,-2));
       
       tVisible = new ArrayList();
       
       bc = new NumericBadConsequence("Pierdes tus tesoros visibles. Jajaja.",0,BadConsequence.MAXTREASURES,0);
       prize = new Prize(2,1);
       unusedMonsters.add(new Monster("Testigos Oculares",6,bc,prize,+2));
       
       bc = new DeathBadConsequence("Hoy no es tu día de suerte. Mueres.");
       prize = new Prize(2,5);
       unusedMonsters.add(new Monster("El gran chulhu",20,bc,prize,+4));
       
       bc = new NumericBadConsequence("Tu gobierno te recorta 2 niveles.",2,0,0);
       prize = new Prize(2,1);
       unusedMonsters.add(new Monster("Serpiente Político",8,bc,prize,-2));
       
       tVisible.add(TreasureKind.ARMOR);
       tVisible.add(TreasureKind.HELMET);
       tHidden.add(TreasureKind.BOTHHANDS);
       bc = new SpecificBadConsequence("Pierdes tu casco y tu armadura visible. Pierdes tus manos ocultas.",0,tVisible,tHidden);
       prize = new Prize(1,1);
       unusedMonsters.add(new Monster("Felpuggoth",2,bc,prize,+5));
       
       tVisible = new ArrayList();
       tHidden = new ArrayList(); 
       
       bc = new NumericBadConsequence("Pierdes 2 niveles.",2,0,0);
       prize = new Prize(4,2);
       unusedMonsters.add(new Monster("Shoggoth",16,bc,prize,-4));
       
       bc = new NumericBadConsequence("Pintalabios negro. Pierdes 2 niveles.",2,0,0);
       prize = new Prize(1,1);
       unusedMonsters.add(new Monster("Lolitagooth",2,bc,prize,+3));
    }
    
    // Iterar sobre cada elemento del array, cambiandolo de sitio a una posición aleatoria. 
    private void shuffleTreasures(){
        Random random = new Random();
        for(int i = 0; i < unusedTreasures.size(); i++){
            
            int pos_nueva = random.nextInt(unusedTreasures.size());
            Treasure aux = unusedTreasures.get(pos_nueva);
            unusedTreasures.set(pos_nueva, unusedTreasures.get(i));
            unusedTreasures.set(i, aux);
        }       
    }
    
    private void shuffleCultists(){
        Random random = new Random();
        for(int i = 0; i < unusedCultists.size(); i++){
            
            int pos_nueva = random.nextInt(unusedCultists.size());
            Cultist aux = unusedCultists.get(pos_nueva);
            unusedCultists.set(pos_nueva, unusedCultists.get(i));
            unusedCultists.set(i, aux);
        }      
    }
    
    private void shuffleMonsters() {
        Random random = new Random();
        for (int i = 0; i < unusedMonsters.size(); i++) {

            int pos_nueva = random.nextInt(unusedMonsters.size());
            Monster aux = unusedMonsters.get(pos_nueva);
            unusedMonsters.set(pos_nueva, unusedMonsters.get(i));
            unusedMonsters.set(i, aux);
        }
    }
    
    public Cultist nextCultist(){
        if(unusedCultists.isEmpty()){
            ArrayList<Cultist> punteroAux = unusedCultists;
            unusedCultists = usedCultists;
            usedCultists = punteroAux;
            shuffleTreasures();
        }
        return unusedCultists.remove(0);
    }
    
    public Treasure nextTreasure(){
        if(unusedTreasures.isEmpty()){
            ArrayList<Treasure> punteroAux = unusedTreasures;
            unusedTreasures = usedTreasures;
            usedTreasures = punteroAux;
            shuffleTreasures();
        }
        return unusedTreasures.remove(0);
    }
    
    public Monster nextMonster(){
        if(unusedMonsters.isEmpty()){
            ArrayList<Monster> punteroAux = unusedMonsters;
            unusedMonsters = usedMonsters;
            usedMonsters = punteroAux;
            shuffleMonsters();
        }
        return unusedMonsters.remove(0);
    }
    
    // ¿Lo borro de unused?
    public void giveTreasureBack(Treasure t){
        usedTreasures.add(t);
        unusedTreasures.remove(t);
    }

    // ¿Lo borro de unused?
    public void giveMonsterBack(Monster m){
        usedMonsters.add(m);
        unusedMonsters.remove(m);
    }
    
    public void initCards(){
        initTreasureCardDeck();
        initMonsterCardDeck();
        initCultistCardDeck();
    }
    
}
